### PR TITLE
Handle `single` distributions in `fANOVA` evaluator

### LINKED ISSUE #2542: `ZeroDivisionError` when plotting parameter importance
Plotting parameter importance of a multi-objective study causes a `ZeroDivisionError: Weights sum to zero, can't be normalized.` The error happens with a with two-objective study, which otherwise runs nicely and produces a Pareto plot without any problem.

## Expected behavior

Generate a plot of parameter importance based on a multi-objective study, as described in the documentation.

## Environment

- Optuna version: 2.6.0
- Python version: 3.8
- OS: Ubuntu 18.04

## Error messages, stack traces, or logs

```
ERROR Traceback (most recent call last):
  File "/opt/mt/anaconda3/envs/develop/lib/python3.8/site-packages/optuna/visualization/_param_importances.py", line 126, in plot_param_importances
    importances = optuna.importance.get_param_importances(
  File "/opt/mt/anaconda3/envs/develop/lib/python3.8/site-packages/optuna/importance/__init__.py", line 91, in get_param_importances
    return evaluator.evaluate(study, params=params, target=target)
  File "/opt/mt/anaconda3/envs/develop/lib/python3.8/site-packages/optuna/importance/_fanova/_evaluator.py", line 121, in evaluate
    evaluator.fit(
  File "/opt/mt/anaconda3/envs/develop/lib/python3.8/site-packages/optuna/importance/_fanova/_fanova.py", line 74, in fit
    self._trees = [_FanovaTree(e.tree_, search_spaces) for e in self._forest.estimators_]
  File "/opt/mt/anaconda3/envs/develop/lib/python3.8/site-packages/optuna/importance/_fanova/_fanova.py", line 74, in <listcomp>
    self._trees = [_FanovaTree(e.tree_, search_spaces) for e in self._forest.estimators_]
  File "/opt/mt/anaconda3/envs/develop/lib/python3.8/site-packages/optuna/importance/_fanova/_tree.py", line 23, in __init__
    statistics = self._precompute_statistics()
  File "/opt/mt/anaconda3/envs/develop/lib/python3.8/site-packages/optuna/importance/_fanova/_tree.py", line 185, in _precompute_statistics
    value = numpy.average(child_values, weights=child_weights)
  File "<__array_function__ internals>", line 5, in average
  File "/opt/mt/anaconda3/envs/develop/lib/python3.8/site-packages/numpy/lib/function_base.py", line 409, in average
    raise ZeroDivisionError(
ZeroDivisionError: Weights sum to zero, can't be normalized

---------------------------------------------------------------------------
ZeroDivisionError                         Traceback (most recent call last)
<ipython-input-10-380c9d9887b3> in <module>
/opt/mt/anaconda3/envs/develop/lib/python3.8/site-packages/... in create_param_importance_plot(self)
    168 
    169     def create_param_importance_plot(self):
--> 170         param_importance_fig = optuna.visualization.plot_param_importances(
    171             self.study, target_name=Q_METRIC_NAME, target=lambda t: t.values[0]
    172         )

/opt/mt/anaconda3/envs/develop/lib/python3.8/site-packages/optuna/visualization/_param_importances.py in plot_param_importances(study, evaluator, params, target, target_name)
    124         return go.Figure(data=[], layout=layout)
    125 
--> 126     importances = optuna.importance.get_param_importances(
    127         study, evaluator=evaluator, params=params, target=target
    128     )

/opt/mt/anaconda3/envs/develop/lib/python3.8/site-packages/optuna/importance/__init__.py in get_param_importances(study, evaluator, params, target)
     89         raise TypeError("Evaluator must be a subclass of BaseImportanceEvaluator.")
     90 
---> 91     return evaluator.evaluate(study, params=params, target=target)

/opt/mt/anaconda3/envs/develop/lib/python3.8/site-packages/optuna/importance/_fanova/_evaluator.py in evaluate(self, study, params, target)
    119 
    120         evaluator = self._evaluator
--> 121         evaluator.fit(
    122             X=trans_params,
    123             y=trans_values,

/opt/mt/anaconda3/envs/develop/lib/python3.8/site-packages/optuna/importance/_fanova/_fanova.py in fit(self, X, y, search_spaces, column_to_encoded_columns)
     72         self._forest.fit(X, y)
     73 
---> 74         self._trees = [_FanovaTree(e.tree_, search_spaces) for e in self._forest.estimators_]
     75         self._column_to_encoded_columns = column_to_encoded_columns
     76         self._variances = {}

/opt/mt/anaconda3/envs/develop/lib/python3.8/site-packages/optuna/importance/_fanova/_fanova.py in <listcomp>(.0)
     72         self._forest.fit(X, y)
     73 
---> 74         self._trees = [_FanovaTree(e.tree_, search_spaces) for e in self._forest.estimators_]
     75         self._column_to_encoded_columns = column_to_encoded_columns
     76         self._variances = {}

/opt/mt/anaconda3/envs/develop/lib/python3.8/site-packages/optuna/importance/_fanova/_tree.py in __init__(self, tree, search_spaces)
     21         self._search_spaces = search_spaces
     22 
---> 23         statistics = self._precompute_statistics()
     24         split_midpoints, split_sizes = self._precompute_split_midpoints_and_sizes()
     25         subtree_active_features = self._precompute_subtree_active_features()

/opt/mt/anaconda3/envs/develop/lib/python3.8/site-packages/optuna/importance/_fanova/_tree.py in _precompute_statistics(self)
    183                     child_values.append(statistics[child_node_index, 0])
    184                     child_weights.append(statistics[child_node_index, 1])
--> 185                 value = numpy.average(child_values, weights=child_weights)
    186                 weight = numpy.sum(child_weights)
    187                 statistics[node_index] = [value, weight]

<__array_function__ internals> in average(*args, **kwargs)

/opt/mt/anaconda3/envs/develop/lib/python3.8/site-packages/numpy/lib/function_base.py in average(a, axis, weights, returned)
    407         scl = wgt.sum(axis=axis, dtype=result_dtype)
    408         if np.any(scl == 0.0):
--> 409             raise ZeroDivisionError(
    410                 "Weights sum to zero, can't be normalized")
    411 

ZeroDivisionError: Weights sum to zero, can't be normalized
```

## Steps to reproduce

1. Configure and run a default study with two objectives (maximize, maximize), the main function returning the result of a rounded float for one objective and an integer division for the second objective.
2. Generate a parameter importance plot with:
```python
fig = optuna.visualization.plot_param_importances(study, target_name=METRIC_NAME, target=lambda t: t.values[0]) 
```

## Additional context (optional)

Apart from the param importances not being generated, the study runs smoothly and I can generate a Pareto front plot without any problem.

### PR DESCRIPTION
**Pull request by xadrianzetx**
## Motivation

This PR fixes #2542 by making `FanovaImportanceEvaluator` correctly handle param distributions with equal lower and upper bounds. This is an alternative implementation, as discussed in #3074 (thanks @hvy!).

## Description of the changes

* Handle `single` distributions in `FanovaImportanceEvaluator.evaluate`
* Write tests

### CODE REVIEW

**Review by hvy — COMMENTED** [review]
Thanks a lot for the quick PR. This looks like a simple solution. One nitpick or suggestion would be to leave a comment around `zero_importances` that `_Fanova` does not expect/support parameter search spaces with size 0.

**Comment by xadrianzetx** [comment]
> This looks like a simple solution.

Yeah, I prefer this implementation as well.

> One nitpick or suggestion would be to leave a comment around zero_importances that _Fanova does not expect/support parameter search spaces with size 0.

Done in 4a69255262ee64eb1c3a71bbdbf998789009bf5a :)

File: tests/importance_tests/test_fanova.py
Comment (hvy) [inline tests/importance_tests/test_fanova.py:89]: Leaving this test might be fine but do you think it's better suited in https://github.com/optuna/optuna/blob/master/tests/importance_tests/test_init.py ? That is, in the general importance tests, not just fANOVA (although the change is limited to this evaluator).

Comment (xadrianzetx) [reply tests/importance_tests/test_fanova.py:89]: I was thinking about it, but concluded that such test would require to also check the behavior of `MeanDecreaseImpurityImportanceEvaluator`, which i think is out of scope here. How about i tackle this in followup? If it works I'll just add parametrized tests as you suggested, and if not I'll also fix the other evaluator.

Comment (hvy) [reply tests/importance_tests/test_fanova.py:89]: That sounds like a much better approach. Let me approve this PR as I think it looks good in that case. 👍 
(I think you can start checking the `MeanDecreaseImpurityImportanceEvaluator` while waiting for a second review, if you like of course)

**Review by hvy — APPROVED** [review]
Thanks a lot for the quick fix, the discussions and suggestion to work on the follow up (https://github.com/optuna/optuna/pull/3085#discussion_r744920304). LGTM!

File: tests/importance_tests/test_fanova.py
Comment (hvy) [reply tests/importance_tests/test_fanova.py:89]: I'm fine either way.

The test you added is obviously reasonable since it's testing the behavior of public API, but at the same time, we've left most of the general behavioral checks to `test_init.py`, given that `get_param_importances` is basically just calling `BaseImportanceEvaluator.evaluate`. With that said, for consistency and simplicity, I slightly lean towards removing.

**Comment by xadrianzetx** [comment]
Btw, @hvy  I've checked `MeanDecreaseImpurityImportanceEvaluator` and it works with this edge case without a problem, so i'll be only better test coverage for the followup :)

File: tests/importance_tests/test_fanova.py
Comment (toshihikoyanase) [inline tests/importance_tests/test_fanova.py:88]: In addition to checking the value of `y`, how about checking the existence of `x` to show a developer-friendly error message?

When developers change the line `importances = {**importances, **zero_importances}`, they may unintendedly drop some of the parameters. For example, if I (wrongly) changed the line to `importances = zero_importances`, this test passed and the other tests failed due to the assertion error of `OrderedDict() != OrderedDict()`. It may not be intuitive, and I suggest the check of `x`.

Moreover, if the importance of `y` is zero, the importance of `x` is automatically one and we can assert `importances["x"] == 1.0`.

**Review by toshihikoyanase — COMMENTED** [review]
Thank you for your PR! The change looks great, but I just added a small suggestion to the test case.

File: tests/importance_tests/test_fanova.py
Comment (xadrianzetx) [reply tests/importance_tests/test_fanova.py:88]: Sorry for small delay! Checking all parameters in this test makes perfect sense to me. Added in a5e4722c7ffbefc23ed6551645c25e021f2862cb.

Comment (toshihikoyanase) [reply tests/importance_tests/test_fanova.py:88]: Great! Thank you for your update.

**Review by toshihikoyanase — APPROVED** [review]
Looks great to me! Thank you for your contribution.

We may have a possible follow-up about the unit test (https://github.com/optuna/optuna/pull/3085#discussion_r744920304), so let's discuss it in a new PR.

**Pull request merged**
